lib/app_profiler.rb

```diff
@@ -60,17 +63,73 @@ module Viewer
   mattr_reader :after_process_queue, default: nil
 
   class << self
-    def run(*args, &block)
-      Profiler.run(*args, &block)
+    def run(*args, with_backend: nil, **kwargs, &block)
+      orig_backend = backend
+      begin
+        self.backend = with_backend if with_backend
+        profiler.run(*args, **kwargs, &block)
+      rescue BackendError
+        yield
+      end
+    ensure
+      AppProfiler.backend = orig_backend
     end
 
     def start(*args)
-      Profiler.start(*args)
+      profiler.start(*args)
     end
 
     def stop
-      Profiler.stop
-      Profiler.results
+      profiler.stop
+      profiler.results
+    end
+
+    def running?
+      @backend&.running?
+    end
+
+    def profiler
+      @backend ||= backend.new
+    end
+
+    def backend=(new_backend)
+      new_profiler_backend = if new_backend.is_a?(String)
```

Comment:
a bit weird i know, but when we have the middleware params we only have a string. I'd rather do the lookup here.

It also feels silly to require that other calls send the string via ::NAME, so i thought may as well take advantage of ruby duck typing here and let it accept either, and resolve it to the type it needs.